pkg/cmd/operator/cmd.go

```diff
@@ -99,6 +105,70 @@ func NewOperator() *cobra.Command {
 				// logger to do nothing.
 				ctrlruntimelog.SetLogger(logr.New(ctrlruntimelog.NullLogSink{}))
 
+				log.Info("checking prerequisites")
+				featureGates, err := platform.GetFeatureGates(ctx)
+				if err != nil {
+					log.WithError(err).Fatal("unable to read feature gates")
+				}
+				awsSecurityTokenServiceGateEnabled := featureGates.Enabled(configv1.FeatureGateAWSSecurityTokenService)
+
+				kubeconfigCommandLinePath := cmd.PersistentFlags().Lookup("kubeconfig").Value.String()
+				rules := clientcmd.NewDefaultClientConfigLoadingRules()
+				rules.ExplicitPath = kubeconfigCommandLinePath
+				kubeconfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(rules, &clientcmd.ConfigOverrides{})
+				cfg, err := kubeconfig.ClientConfig()
+				if err != nil {
+					log.WithError(err).Fatal("failed to parse kubeconfig")
+				}
+				coreClient, err := corev1client.NewForConfig(cfg)
+				if err != nil {
+					log.WithError(err).Fatal("failed to set up client")
+				}
+
+				// The status quo for this controller is to LIST + WATCH all Secrets on the cluster. This consumes
+				// more resources than necessary on clusters where users put other data in Secrets themselves, as we
+				// hold that data in our cache and never do anything with it. The reconcilers mainly need to react to
+				// changes in Secrets created for CredentialRequests, which they control and can label, allowing us
+				// to filter the LIST + WATCH down and hold the minimal set of data in memory. However, two caveats:
+				// - in passthrough and mint mode, admin credentials are also provided by the user through Secrets,
+				//   and we need to watch those, but we can't label them
+				// - on existing clusters, secrets exist from previous work these reconcilers did that will not have
+				//   Secrets labelled
+				// We could solve the second issue with an interim release of this controller that labels all previous
+				// Secrets, but does not restrict the watch stream.
```

Comment:
> did you look into just always labeling all the secrets at startup time, before going into list/watch mode?

It was not clear to me how to do this without a race / with any guarantee that this process finishes - we would want to make sure that no Secrets are created or mutated between the moment we are done labelling and when we start the LIST+WATCH. Thoughts?

> and/or recording some sort of "labeling complete" indicator in the ClusterOperator status so we can avoid doing it every time we start?

Similarly here, we'd ideally want to keep that field on the OperatorStatus up to date over time, right, so we'd need to know, given a watch even we're processing in the controller at runtime, whether or not we need to update the OperatorStatus. It's not clear how we'd know when we were processing the last un-labelled Secret and therefore could call the job done.

Reply:
> we would want to make sure that no Secrets are created or mutated 

secret creation doesn't matter, right?  because there are two things creating secrets:

1) us - no worries, label will be applied.  also we won't be creating any until after we get through the initialization phase
2a) someone else, but the secret isn't ref'd by a credreq - don't care
2b) someone else, they create a credreq and a secret ref'd by the cred req - we'll pick up this secret when we see the credreq object and go to inject content into the secret, at which point we can label it.

so i think creation is probably handled.

Mutation isn't, i think, a concern, because mutation implies the secret already existed when we were initalizing, in which case we'd have already labeled it.  I guess the one gap might be:

1) we list + label secrets
2) user edits secret and stomps the label away
3) we never reconcile that secret

though in theory that could also be a concern at runtime?  If someone edits a secret to remove our label, will we get notified of the change to the secret, given that it no longer matches our filter?  Certainly if they do it while the controller is down we'd have that problem unless we relist+label all secrets when we start up.

Regardless i don't see how doing this in a previous release helps.  If there is a window between "we listing/watching all secrets and label them" and "we're only watching labeled secrets" in which a secret can be created/mutated such that we never reconcile it again, then that window would exist whether we:
1) ship one release that has a controller looking at all of them, then in a subsequent release we stop doing so
or
2) ship, in a single release, logic that looks at all of them, and then stops doing so

no?

Also the current logic assumes "if we list all secrets and they are all labeled, we're safe to filter", but doesn't that have similar race condition concerns? (a secret could be created/mutated after we finish our listing and think it's safe to filter) At least if we label them all while listing them, assuming we are successful in labeling them, we don't have to wait a release for most clusters to be able to enable the logic.  But that still doesn't address the race concern.

Reply:
Addressed in the latter commits.